Add cy.origin command API page (#4439)
cypress-io/cypress-documentation: content/api/commands/origin.md

---
title: origin
---

Visit multiple domains of different
[origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin)
in a single test.

In normal use, a single Cypress test may only run commands in a single origin, a
limitation determined by standard web security features of the browser. The
`cy.origin()` command allows your tests to bypass this limitation.

<Alert type="warning">

<strong class="alert-header"><Icon name="exclamation-triangle"></Icon>
Experimental</strong>

TODO Maybe rework this & move to a partial

The `origin` API is currently experimental, and can be enabled by setting
the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
to `true` in the Cypress config or by

> using [`Cypress.config()`](/api/cypress-api/config) at the top of a spec file.

Reviewer: The new flag (experimentalSessionAndOrigin) is only configurable globally and can't be set in a spec file.

Enabling this flag does the following:

- It adds the [`cy.session()`](/api/commands/session) and `cy.origin()`
  commands, and [`Cypress.session`](/api/cypress-api/session) API.
- It adds the following new behaviors (that will be the default in a future
  major update of Cypress) at the beginning of each test:
  - The page is cleared (by setting it to `about:blank`).
  - All active session data (cookies, `localStorage` and `sessionStorage`)
    across all domains are cleared.
- It overrides

>   the [`Cypress.Cookies.preserveOnce()`](/api/cypress-api/cookies#Preserve-Once) and [`Cypress.Cookies.defaults()`](/api/cypress-api/cookies#Defaults) methods.

Author: I actually don't know what this means either! It's copied from the `cy.session` docs: https://docs.cypress.io/api/commands/session

>   the [`Cypress.Cookies.preserveOnce()`](/api/cypress-api/cookies#Preserve-Once) and [`Cypress.Cookies.defaults()`](/api/cypress-api/cookies#Defaults) methods.

Reviewer 2: I guess it's fine to leave. I'd assume this is accurate, but I don't remember this.

>   the [`Cypress.Cookies.preserveOnce()`](/api/cypress-api/cookies#Preserve-Once) and [`Cypress.Cookies.defaults()`](/api/cypress-api/cookies#Defaults) methods.

Author: I'm going to leave this for now and maybe revisit later

- Cross-domain requests will no longer fail immediately, but instead, time out
  based on [`pageLoadTimeout`](/guides/references/configuration#Timeouts).
- Tests will no longer wait on page loads before moving on to the next test.

Because the page is cleared before each
test, [`cy.visit()`](/api/commands/visit) must be explicitly called in each test
to visit a page in your application.

</Alert>

## Syntax

```js
cy.origin(url, callbackFn)
cy.origin(url, options, callbackFn)
```

### Usage

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```js
const hits = getHits() // Defined elsewhere
// Run commands in a secondary origin, passing in any serializable values we need
cy.origin('https://www.acme.com', { args: { hits } }, ({ hits }) => {
  // Inside callback baseUrl is https://www.acme.com
  cy.visit('/history/founder')
  // Commands are executed in secondary origin
  cy.get('h1').contains('About our Founder, Marvin Acme')
  // Passed in values are accessed via callback args
  cy.get('#hitcounter').contains(hits)
})
// Even though we are outside the secondary origin block, we are still on acme.com
// so return to baseUrl

> cy.visit('/')

Reviewer 3: I did not realize this would work and i had to test it it, it does totally work!

// Continue running commands on primary origin
cy.get('h1').contains('My cool site under test')
```

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

```js

> const hits = 9000

Reviewer 3: `hits` above is the result of a function, but here is a constant, we should be consistent.

Author: Very good point, changing it to a `require`d value

// This should be inside the callback, Cypress needs to be injected before visit

> cy.visit('https://www.acme.com/history/founder')

Reviewer 3: IMO we should investigate what it would take to make this work for GA.

Author: I do think people will find this quite confusing

cy.origin('https://www.acme.com', () => {

>   // Won't work because Cypress is not present on secondary origin

Author: Replacing with `// Fails because origin was visited before cy.origin() block`, let me know if this is OK

  cy.get('h1').contains('About our Founder, Marvin Acme')
  // Won't work because hits is not passed in via args
  cy.get('#hitcounter').contains(hits)
})
// Won't work because still on acme.com
cy.get('h1').contains('My cool site under test')
```

### Arguments

**<Icon name="angle-right"></Icon> url** **_(String)_**

A URL specifying the secondary origin in which the callback is to be executed.
This should at the very least contain a hostname, and may also include the
protocol, port number & path. This argument will be used in two ways.

Firstly, it uniquely identifies a secondary origin in which the commands in the
callback will be executed. The test-runner will inject the Cypress runtime into
this origin, and then send it code to evaluate in that origin, without violating
the browser's same-origin policy.

Secondly, it temporarily overrides the `baseUrl` configured in your
[global configuration](/guides/references/configuration#Global) whilst inside
the callback. So `cy.visit()` will navigate relative to this URL, not the
configured `baseUrl`.

> 

Reviewer 3: I'd remove 'temporarily' from the second point.

**<Icon name="angle-right"></Icon> options** **_(Object)_**

Pass in an options object to control the behavior of `cy.origin()`.

| option | description                                                                                                                                                                                                    |
| ------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| args   | Plain JavaScript object which will be serialized and sent from the primary origin to the secondary origin, where it will be deserialized and passed into the callback function as its first and only argument. |

<Alert type="warning">

> The `args` object is the **only** mechanism via which data may be injected into
> the callback, the callback is **not** a closure and does not retain access to
> the JavaScript context in which it was declared.

Reviewer 3: ```suggestion
The `args` object is the **only** mechanism via which data may be injected into
the callback, the callback is **not** a closure and does not retain access to
the JavaScript context in which it was declared. Values passed into `args`
**must** be serializable.
```

</Alert>

**<Icon name="angle-right"></Icon> callbackFn** **_(Function)_**

The function containing the commands to be executed in the secondary origin.

This function will be stringified, sent to the Cypress instance in the secondary
origin and evaluated. If the `args` option is specified, the deserialized args
object will be passed into the function as its first and only argument.

There are a number of limitations placed on commands run inside the callback,
please see [Callback restrictions](#Callback-restrictions) section below for a
full list.

> The function's return value is ignored. ???

Reviewer: `cy.origin` can be chained further as we honor the function's return value as long as it is serializable.

Reviewer 3: Specifically we return either the result of the callback, if no cy commands are present or the result of the last cy command.

### Yields [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management)

- `cy.origin()` yields `null`.

> - `cy.origin()` cannot be chained further.

Reviewer 4: `cy.origin()` yields the value returned from the callback, not `null`: https://github.com/cypress-io/cypress/pull/18075/files#r852264190

## Examples

### Using dynamic data in a secondary origin

Callbacks are executed inside an entirely separate instance of Cypress, so
arguments must be transmitted to the other instance by means of serialization &

> deserialization. The interface for this mechanism is the `args` option.

Reviewer 3: Specifically we use the structured clone algorithm, if you want to delve that deep.

https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm

```js

> const sentArgs = { password: 'P@55w0rd!!!1!?!111one' }
> cy.origin(
>   'supersecurelogons.com',
>   // Send the args here...
>   { args: sentArgs },
>   // ...and receive them at the other end here!
>   (receivedArgs) => {
>     const { password } = receivedArgs
>     cy.visit('/login')
>     cy.get('input#password').type(password)
>     cy.contains('button', 'Login').click()
>   }
> )

Reviewer 3: Seems more normal to send username and password. I'd also destructure `receivedArgs` earlier, but i assume you wrote it like you did to drive home the point that the args are received as params. so either way is probably fine.

```suggestion
const sentArgs = { username: 'username', password: 'P@55w0rd!!!1!?!111one' }
cy.origin(
  'supersecurelogons.com',
  // Send the args here...
  { args: sentArgs },
  // ...and receive them at the other end here!
  ({ username, password }) => {
    cy.visit('/login')
    cy.get('input#username').type(username)
    cy.get('input#password').type(password)
    cy.contains('button', 'Login').click()
  }
)
```

Author: Exactly correct!

```

Note: You can just replace `sentArgs` and `receivedArgs` with `args` if you
want, the naming in this example is purely for clarity.

### Navigating to secondary origin with cy.visit

When navigating to a secondary origin using `cy.visit()`, it is essential to
trigger the navigation **after** entering the origin callback, otherwise a
cross-domain error will be thrown.

```js
// Do things in primary domain...

> cy.origin('acme.com', () => {

Reviewer 3: ```suggestion
cy.origin('https://www.acme.com/', () => {
```

>   // Visit https://www.acme.com/history/founder

Reviewer: The subdomain won't be added since it wasn't passed to `cy.origin`. In this case, the relative url would expand to: `https://acme.com/history/founder`.

  cy.visit('/history/founder')
  cy.get('h1').contains('About our Founder, Marvin Acme')
})
```

```js
// TODO Example with onBeforeLoad...
```

TODO baseUrl

TODO default protocol

### Navigating to secondary origin with UI

When navigating to a secondary origin by clicking a link or button in the
primary origin, it is essential to trigger the navigation _before_ entering the
origin callback, otherwise a cross-domain error will be thrown.

```js

> // Click button in primary origin that navigates to https://acme.com

Reviewer 2: Reducing line width to avoid scrolling

```suggestion
// Button in primary origin goes to https://acme.com
```

cy.contains('button', 'Go to Acme.com').click()

cy.origin('acme.com', () => {
  // No cy.visit is needed as the button brought us here
  cy.get('h1').contains('ACME CORP')
})
```

### Navigating to multiple secondary origins in succession

Callbacks may **not** themselves contain `cy.origin()` calls, so when visiting

> multiple origins, do so at the top level of the test.

Reviewer 3: Do you want to mention that we do plan to support this in the GA release?

```js
cy.origin('foo.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'foo.com')
})

cy.origin('bar.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'bar.com')
})

cy.origin('baz.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'baz.com')
})
```

TODO Stabilization?

### Waiting to return to primary origin

Sometimes, a secondary origin returns to the primary origin as a result of user
action (for example, by clicking a "Login" button on a syndicated login
provider). In this situation, your test should wait for the navigation to
complete before making further assertions, otherwise a cross-domain error will
be thrown.

```js
cy.visit('/home')
// This will take us to the secondary origin
cy.contains('button', 'Go to someothersite').click()

cy.origin('someothersite.com', () => {
  // Click button that takes us back to primary origin
  cy.contains('button', 'Go back').click()
)

> // Wait until confirmation we are back at the primary origin before continuing

Author: OK, I'm going to delete the whole "Waiting to return to primary origin" section then. I thought this was necessary for some reason, but I can't remember why any more...

do this with `cy.origin()`.

```js
Cypress.Commands.add('login', (username, password) => {
  // Remember to pass in dependencies via `args`
  const args = { username, password }
  cy.origin('auth-provider.com', { args }, ({ username, password }) => {
    // Go to https://auth-provider.com/login
    cy.visit('/login')
    cy.contains('Username').find('input').type(username)
    cy.contains('Password').find('input').type(password)
    cy.get('button').contains('Login').click()
  })
  // Wait until confirmation we are back at the primary origin before continuing
  cy.url().should('contain', '/home')
})
```

However, having to go through an entire login flow before every test is not very
performant. Up until now you could get around this by putting login code in the
first test of your file, then performing subsequent tests reusing the same
session. However, once the `experimentalSessionAndOrigin` flag is activated this
is no longer possible, as all session state is now cleared between tests. So to
avoid this overhead we have added the [`cy.session()`](/api/commands/session)
command, which allows you to easily cache session information and reuse it

> across not just tests, but test files too. So now let's enhance our custom login

Reviewer 2: session info is NOT available across test files. It resets across test files. https://github.com/cypress-io/cypress/issues/17710

```suggestion
across tests. So now let's enhance our custom login
```

Author: D'oh!

command with `cy.session()` for a complete syndicated login flow with session
caching and validation. No mocking, no workarounds, no third-party plugins!

```js
Cypress.Commands.add('login', (username, password) => {
  const args = { username, password }
  cy.session(

>     // The username & password combination can be used as the cache key too

Reviewer 2: ```suggestion
    // Username & password can be used as the cache key too
```

    args,
    () => {

>       cy.origin('auth-provider.com', { args }, ({ username, password }) => {

Reviewer 2: Reducing line width

```suggestion
      cy.origin('my-auth.com', { args }, ({ username, password }) => {
```

[discouraged](/guides/references/best-practices#Having-tests-rely-on-the-state-of-previous-tests)
we know some users have historically written tests this way, often to get around
the same-origin restrictions. But with `cy.origin()` you no longer need these
kind of brittle hacks, as your multi-origin logic can all reside in a single
test, like the following.

<Badge type="success">After</Badge> One big test using `cy.origin()`

```js
it('securely edits content', () => {
  cy.origin('supersecurelogons.com', () => {
    cy.visit('https"//supersecurelogons.com')
    cy.get('input#password').type('Password123!')
    cy.get('button#submit').click()
  })

  cy.origin('mycms.com', () => {
    cy.url().should('contain', 'cms')
    cy.get('#current-user').contains('logged in')
    cy.get('button#edit-1').click()
    cy.get('input#title').type('Updated title')
    cy.get('button#submit').click()
    cy.get('.toast').type('Changes saved!')
  })

  cy.visit('/items/1')
  cy.get('h1').contains('Updated title')
})
```

Always remember,
[Cypress tests are not unit tests](https://docs.cypress.io/guides/references/best-practices#Creating-tiny-tests-with-a-single-assertion).

### Serialization

> When entering a `cy.origin()` block, the test-runner injects the Cypress
> runtime, with all your configurations settings, into the requested origin, and

Reviewer 2: ```suggestion
When entering a `cy.origin()` block, Cypress injects itself at
runtime, with all your configurations settings, into the requested origin, and
```

sets up bidirectional communication with that instance. This coordination model
requires that any data sent from one instance to another be
[serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization)
for transmission. It is very important to understand that variables **inside**
the callback are not shared with the scope **outside** the callback. For example
this will not work:

```js
const foo = 1
cy.origin('somesite.com', () => {
  cy.visit('/')
  // This line will throw a ReferenceError because `foo` is not defined in the
  // scope of the callback
  cy.get('input').type(foo)
})

> ```

Reviewer 2: Usually when providing examples that don't work we need big ❗ icons. People...will.....copy....this...code! And not read the actual text that this is a non-working example. 😅  Example: https://docs.cypress.io/guides/core-concepts/introduction-to-cypress#Commands-Are-Asynchronous

cy.origin('https://www.acme.com', { args: { hits } }, ({ hits }) => {
  // Inside callback baseUrl is https://www.acme.com
  cy.visit('/history/founder')
  // Commands are executed in secondary origin
  cy.get('h1').contains('About our Founder, Marvin Acme')
  // Passed in values are accessed via callback args
  cy.get('#hitcounter').contains(hits)
})
// Even though we're outside the secondary origin block,
// we're still on acme.com so return to baseUrl
cy.visit('/')
// Continue running commands on primary origin
cy.get('h1').contains('My cool site under test')
```

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

```js
const hits = getHits()
// cy.visit() should be inside cy.origin() callback
cy.visit('https://www.acme.com/history/founder')
cy.origin('https://www.acme.com', () => {
  // Fails because origin was visited before cy.origin() block
  cy.get('h1').contains('About our Founder, Marvin Acme')
  // Fails because hits is not passed in via args
  cy.get('#hitcounter').contains(hits)
})
// Won't work because still on acme.com
cy.get('h1').contains('My cool site under test')
```

### Arguments

**<Icon name="angle-right"></Icon> url** **_(String)_**

> A URL specifying the secondary origin in which the callback is to be executed.

Reviewer: Do we want to explicitly state query params are not supported?

This should at the very least contain a hostname, and may also include the
protocol, port number & path.

This argument will be used in two ways:

1. It uniquely identifies a secondary origin in which the commands in the
   callback will be executed. Cypress will inject itself into this origin, and
   then send it code to evaluate in that origin, without violating the browser's
   same-origin policy.

2. It overrides the `baseUrl` configured in your
   [global configuration](/guides/references/configuration#Global) while inside

>    the callback. So `cy.visit()` will navigate relative to this URL, not the

Reviewer: and `cy.request()`

please see [Callback restrictions](#Callback-restrictions) section below for a
full list.

### Yields [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management)

- `cy.origin()` yields the value yielded by the last Cypress command in the
  callback function.
- If the callback contains no Cypress commands, `cy.origin()` yields the return
  value of the function.

## Examples

### Using dynamic data in a secondary origin

Callbacks are executed inside an entirely separate instance of Cypress, so
arguments must be transmitted to the other instance by means of
[the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm).
The interface for this mechanism is the `args` option.

```js
const sentArgs = { username: 'username', password: 'P@55w0rd!' }
cy.origin(
  'supersecurelogons.com',
  // Send the args here...
  { args: sentArgs },
  // ...and receive them at the other end here!
  ({ username, password }) => {
    cy.visit('/login')
    cy.get('input#username').type(username)
    cy.get('input#password').type(password)
    cy.contains('button', 'Login').click()
  }
)
```

> Note: You can just replace `sentArgs` and `receivedArgs` with `args` if you
> want, the naming in this example is purely for clarity.

Reviewer 3: ```suggestion
Note: You can just replace `sentArgs` with `args` if you
want, the naming in this example is purely for clarity.
```

### Navigating to secondary origin with cy.visit

When navigating to a secondary origin using `cy.visit()`, it is essential to
trigger the navigation **after** entering the origin callback, otherwise a
cross-domain error will be thrown.

```js
// Do things in primary domain...

cy.origin('https://www.acme.com', () => {
  // Visit https://www.acme.com/history/founder
  cy.visit('/history/founder')
  cy.get('h1').contains('About our Founder, Marvin Acme')
})
```

> TODO baseUrl / default protocol

Reviewer 3: Still outstanding? if not remove

cy.contains('button', 'Go').click()

cy.origin('www.acme.com', () => {
  // No cy.visit is needed as the button brought us here
  cy.get('h1').contains('ACME CORP')
})
```

### Navigating to multiple secondary origins in succession

Callbacks may **not** themselves contain `cy.origin()` calls, so when visiting
multiple origins, do so at the top level of the test.

```js
cy.origin('foo.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'foo.com')
})

cy.origin('bar.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'bar.com')
})

cy.origin('baz.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'baz.com')
})
```

<Alert type="info">

A future version of Cypress will allow the use of nested `cy.origin()` calls.

</Alert>

### SSO login custom command

A very common requirement is logging in to a site before running a test. If
login itself is not the specific focus of the test, it's good to encapsulate
this functionality in a `login`
[custom command](/api/cypress-api/custom-commands) so you don't have to
duplicate this login code in every test. Here's an idealized example of how to
do this with `cy.origin()`.

```js

> Cypress.Commands.add('login', (username, password) => {

Reviewer: I wonder if we should just demonstrate the example using `cy.session()`? I'm worried someone will see this example and copy it, not realizing they should use the below example.

Author: I like the approach of gradually increasing complexity, but I agree that blind copying is a possibility, so I added an "Inefficient Usage" warning & icon

So now let's enhance our custom login command with `cy.session()` for a complete
syndicated login flow with session caching and validation. No mocking, no
workarounds, no third-party plugins!

```js
Cypress.Commands.add('login', (username, password) => {
  const args = { username, password }
  cy.session(
    // Username & password can be used as the cache key too
    args,
    () => {
      cy.origin('my-auth.com', { args }, ({ username, password }) => {
        cy.visit('/login')
        cy.contains('Username').find('input').type(username)
        cy.contains('Password').find('input').type(password)
        cy.get('button').contains('Login').click()
      })
      cy.url().should('contain', '/home')
    },
    {
      validate() {
        cy.request('/api/user').its('status').should('eq', 200)
      },
    }
  )
})
```

## Notes

### Migrating existing tests

Enabling the `experimentalSessionAndOrigin` flag makes the test-runner work
slightly differently, and some test suites that rely on the existing behaviour
may have to be updated. The most important of these changes is **test
isolation**. This means that after every test, the current page is reset to

> [`about:blank`](https://en.wikipedia.org/wiki/About_URI_scheme#Standardization)

Reviewer: Is there a better reference than wikipedia?

Author: I can't find anything helpful, I'm going to remove the link

bidirectional communication with that instance. This coordination model requires
that any data sent from one instance to another be
[serialized](https://developer.mozilla.org/en-US/docs/Glossary/Serialization)
for transmission. It is very important to understand that variables **inside**
the callback are not shared with the scope **outside** the callback. For example
this will not work:

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

```js
const foo = 1
cy.origin('somesite.com', () => {
  cy.visit('/')
  // This line will throw a ReferenceError because `foo` is not defined in the
  // scope of the callback
  cy.get('input').type(foo)
})
```

Instead, the variable must be explicitly passed into the callback using the
`args` option:

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```js
const foo = 1
cy.origin('somesite.com', { args: { foo } }, ({ foo }) => {
  cy.visit('/')
  // Now it will pass
  cy.get('input').type(foo)
})
```

Underneath the hood, Cypress uses
[`JSON.stringify()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)
and

> [`JSON.parse()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse)

Reviewer 3: technically we send the args object as is through `postMessage` which will serialize it with the structured clone algorithm.

https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage

(or equivalent library functions, implementation details may change) to

> serialize and deserialize the `args` object. This introduces a number of

Reviewer: We don't use `JSON.stringify` or `parse`. @AtofStryker and @chrisbreiding, would referencing the structured clone algorithm be sufficient here or are there more limitations we need to specify?

Reviewer 5: It might be good to link [things that don't work with structured clone](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone) here and more-so focus on what isn't serializable vs what is.

> serialize and deserialize the `args` object. This introduces a number of

Reviewer 5: Technically the `args` option can be anything as it is generic. Something like this below also works
```js
    cy.origin('somesite.com', { args: 'my string' }, (shouldBeMyString) => {
      expect(shouldBeMyString).to.equal('my string')
    })
```
```suggestion
serialize and deserialize the `args` option. This introduces a number of
```

restrictions on the data which may be transmitted:

- Primitive values only: object, array, string, number, boolean, null

> - Date objects will be converted to strings

Reviewer 5: Dates are serializable by `structuredClone()`. I wonder if it's worth talking about what isn't serializable vs what is, or just linking users to [things that dont work with structured clone](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone) and covering the basics, like functions, DOM nodes, and complex object are not serializable.
```suggestion
```

Author: OK more types are actually allowed than I realized so I've removed the list altogether & just linked to MDN as you suggested.

> Underneath the hood, Cypress uses
> [`JSON.stringify()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify)
> and
> [`JSON.parse()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse)
> (or equivalent library functions, implementation details may change) to
> serialize and deserialize the `args` object. This introduces a number of
> restrictions on the data which may be transmitted:
> 
> - Primitive values only: object, array, string, number, boolean, null
> - Date objects will be converted to strings
> - Inherited properties are not serialized
> - Circular references are not allowed and will throw an error

Reviewer 5: How does something like this sound?

```suggestion
Underneath the hood, Cypress uses
the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) to
serialize and deserialize the `args` option. This introduces a number of
[restrictions](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone) on the data which may be transmitted.
```

Reviewer: I would remove "Underneath the hood" and just say "The structured clone algorithm is used to..."

Reviewer 5: @mschile updated the suggestion. How does that look?

Author: I updated this section, see what you think

Reviewer 5: @debrisapron much more accurate. Thank you!

### Callback restrictions

Because of the way in which the callback is transmitted and executed, there are
certain limitations on what code may be run inside it. In particular, the
following Cypress commands will throw errors if used in the callback:

- `cy.origin()`
- [`cy.intercept()`](/api/commands/intercept)
- [`cy.session()`](/api/commands/session)
- [`cy.server()`](/api/commands/server)
- [`cy.route()`](/api/commands/route)

> - [`Cypress.Server.defaults()`](/api/cypress-api/cypress-server)

Reviewer: This won't throw an error, we just don't sync big Cypress defaults between origins (e.g. Cypress.Server.defaults(), Cypress.Screenshot.defaults(), etc)

Reviewer 5: @mschile I think we have an exception for [server](https://github.com/cypress-io/cypress/blob/feature-multidomain/packages/driver/src/multi-domain/unsupported_apis.ts#L11).

Author: Sorry, I'm a little confused here - I actually copied this list from the notes of one of the feature meetings & don't understand it entirely. Can you suggest a copy change?

Reviewer: Sorry for the confusion. The error list you have is correct.

If needed, we can also state we don't sync Cypress defaults between origins.

---
title: origin
---

> Visit multiple domains of different
> [origin](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin)
> in a single test.

Reviewer: ```suggestion
Visit multiple 
[origins](https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin)
in a single test.
```

In normal use, a single Cypress test may only run commands in a single origin, a
limitation determined by standard web security features of the browser. The
`cy.origin()` command allows your tests to bypass this limitation.

<Alert type="warning">

<strong class="alert-header"><Icon name="exclamation-triangle"></Icon>
Experimental</strong>

> The `origin` API is currently experimental, and can be enabled by setting

Reviewer 6: should this be `cy.origin()`?

the [`experimentalSessionAndOrigin`](/guides/references/experiments) flag
to `true` in the Cypress config.

Enabling this flag does the following:

- It adds the [`cy.session()`](/api/commands/session) and `cy.origin()`
  commands, and [`Cypress.session`](/api/cypress-api/session) API.
- It adds the following new behaviors (that will be the default in a future

>   major update of Cypress) at the beginning of each test:

Reviewer 6: ```suggestion
  major version release of Cypress) at the beginning of each test:
```

  - The page is cleared (by setting it to `about:blank`).
  - All active session data (cookies, `localStorage` and `sessionStorage`)
    across all domains are cleared.
- It overrides
  the [`Cypress.Cookies.preserveOnce()`](/api/cypress-api/cookies#Preserve-Once) and

>   [`Cypress.Cookies.defaults()`](/api/cypress-api/cookies#Defaults) methods.

Reviewer 6: What does it override to? Do users need to change their tests because of this?

Reviewer 3: Maybe better, maybe not, but this is the second question we have gotten about this line.
```suggestion
- It supersedes
  the [`Cypress.Cookies.preserveOnce()`](/api/cypress-api/cookies#Preserve-Once) and
  [`Cypress.Cookies.defaults()`](/api/cypress-api/cookies#Defaults) methods.
```

> - Cross-domain requests will no longer fail immediately, but instead, time out

Reviewer: ```suggestion
- Cross-origin requests will no longer fail immediately, but instead, time out
```

- Tests will no longer wait on page loads before moving on to the next test.

Because the page is cleared before each
test, [`cy.visit()`](/api/commands/visit) must be explicitly called in each test
to visit a page in your application.

</Alert>

## Syntax

```js
cy.origin(url, callbackFn)
cy.origin(url, options, callbackFn)
```

### Usage

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```js
const hits = getHits() // Defined elsewhere
// Run commands in secondary origin, passing in serializable values
cy.origin('https://www.acme.com', { args: { hits } }, ({ hits }) => {
  // Inside callback baseUrl is https://www.acme.com
  cy.visit('/history/founder')
  // Commands are executed in secondary origin
  cy.get('h1').contains('About our Founder, Marvin Acme')
  // Passed in values are accessed via callback args
  cy.get('#hitcounter').contains(hits)
})
// Even though we're outside the secondary origin block,
// we're still on acme.com so return to baseUrl
cy.visit('/')
// Continue running commands on primary origin
cy.get('h1').contains('My cool site under test')
```

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

> ```js

Reviewer 6: IMO it might be easier to break this into two examples since both commands within the callback are incorrect but technically the second command wold never execute

const hits = getHits()
// cy.visit() should be inside cy.origin() callback
cy.visit('https://www.acme.com/history/founder')
cy.origin('https://www.acme.com', () => {
  // Fails because origin was visited before cy.origin() block
  cy.get('h1').contains('About our Founder, Marvin Acme')
  // Fails because hits is not passed in via args
  cy.get('#hitcounter').contains(hits)
})
// Won't work because still on acme.com
cy.get('h1').contains('My cool site under test')
```

### Arguments

**<Icon name="angle-right"></Icon> url** **_(String)_**

A URL specifying the secondary origin in which the callback is to be executed.
This should at the very least contain a hostname, and may also include the
protocol, port number & path. Query params are not supported.

This argument will be used in two ways:

> 1. It uniquely identifies a secondary origin in which the commands in the

Reviewer 6: should we be using secondary origin or different origin throughout these docs? technically they could visit a t3rd or 4th and this terminology sounds like only a 2nd is supported

Reviewer 7: I agree this could be confusing. We do want to make clear the binary distinction between the primary origin and secondary origins without folks confusing them with being ordinal. I'm not sure if primary/different is a good pairing. Maybe primary/non-primary? 

I feel like we should define these terms somewhere early in this doc, so users have that context as we use them throughout.

   callback will be executed. Cypress will inject itself into this origin, and

>    then send it code to evaluate in that origin, without violating the browser's

Reviewer 7: It's kind of both, but I assume this is meant to be the user's callbackFn. I think "send it code" is correct, but maybe a bit vague...

The `args` object is the **only** mechanism via which data may be injected into
the callback, the callback is **not** a closure and does not retain access to
the JavaScript context in which it was declared. Values passed into `args`
**must** be serializable.

</Alert>

**<Icon name="angle-right"></Icon> callbackFn** **_(Function)_**

The function containing the commands to be executed in the secondary origin.

This function will be stringified, sent to the Cypress instance in the secondary
origin and evaluated. If the `args` option is specified, the deserialized args
object will be passed into the function as its first and only argument.

There are a number of limitations placed on commands run inside the callback,
please see [Callback restrictions](#Callback-restrictions) section below for a
full list.

### Yields [<Icon name="question-circle"/>](/guides/core-concepts/introduction-to-cypress#Subject-Management)

- `cy.origin()` yields the value yielded by the last Cypress command in the
  callback function.
- If the callback contains no Cypress commands, `cy.origin()` yields the return
  value of the function.
- In either of the two cases above, if the value is not serializable,

>   `cy.origin()` yields null.

Reviewer 6: @mjhenkes is this correct or do we throw an error?

Reviewer 7: It's not correct. We don't yield null if the subject is not serializable. But we don't throw immediately either. We only throw if the tests tries to access the unserializable value.

So this is fine:

```js
cy.origin('foobar.com', () => {
  cy.get('#foo') // yields element (not serializable)
})
// .. other commands ...
```

This will throw an error:

```js
cy.origin('foobar.com', () => {
  cy.get('#foo') // yields element (not serializable)
})
.then(($el) => {
  $el.text() // <- throws at this point
})
```

Author: I'm fixing this

## Examples

### Using dynamic data in a secondary origin

Callbacks are executed inside an entirely separate instance of Cypress, so
arguments must be transmitted to the other instance by means of
[the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm).
The interface for this mechanism is the `args` option.

```js
const sentArgs = { username: 'username', password: 'P@55w0rd!' }
cy.origin(
  'supersecurelogons.com',
  // Send the args here...
  { args: sentArgs },
  // ...and receive them at the other end here!
  ({ username, password }) => {
    cy.visit('/login')
    cy.get('input#username').type(username)
    cy.get('input#password').type(password)
    cy.contains('button', 'Login').click()
  }
)
```

Note: You can just replace `sentArgs` with `args` if you want, the naming in

> this example is purely for clarity.

Reviewer 6: I think we can drop this

### Yielding a value

Values returned or yielded from the callback function **must** be serializable
or they will not be returned to the primary origin. For example, the following
will not work:

**<Icon name="exclamation-triangle" color="red"></Icon> Incorrect Usage**

```js
cy.origin('https://www.acme.com', () => {
  cy.visit('/')
  cy.get('h1') // Yields an element, which can't be serialized...
}).contains('ACME CORP') // ...so this will fail
```

Instead, you must explicitly yield a serializable value:

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```js
cy.origin('https://www.acme.com', () => {
  cy.visit('/')
  cy.get('h1').invoke('textContent') // Yields a string...
}).should('equal', 'ACME CORP') // 👍
```

### Navigating to secondary origin with cy.visit

When navigating to a secondary origin using `cy.visit()`, it is essential to
trigger the navigation **after** entering the origin callback, otherwise a

> cross-domain error will be thrown.

Reviewer: ```suggestion
cross-origin error will be thrown.
```

```js

> // Do things in primary domain...

Reviewer 7: ```suggestion
// Do things in primary origin...
```

cy.origin('www.acme.com', () => {
  // Visit https://www.acme.com/history/founder
  cy.visit('/history/founder')
  cy.get('h1').contains('About our Founder, Marvin Acme')
})
```

Here the `baseUrl` inside the callback is set to `www.acme.com` and the protocol
defaults to `https`. When `cy.visit()` is called with the path
`/history/founder`, the three are concatenated to make
`https://www.acme.com/history/founder`.

### Navigating to secondary origin with UI

When navigating to a secondary origin by clicking a link or button in the
primary origin, it is essential to trigger the navigation _before_ entering the

> origin callback, otherwise a cross-domain error will be thrown.

Reviewer: ```suggestion
origin callback, otherwise a cross-origin error will be thrown.
```

```js
// Button in primary origin goes to https://www.acme.com
cy.contains('button', 'Go').click()

cy.origin('www.acme.com', () => {
  // No cy.visit is needed as the button brought us here
  cy.get('h1').contains('ACME CORP')
})
```

### Navigating to multiple secondary origins in succession

Callbacks may **not** themselves contain `cy.origin()` calls, so when visiting
multiple origins, do so at the top level of the test.

```js
cy.origin('foo.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'foo.com')
})

cy.origin('bar.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'bar.com')
})

cy.origin('baz.com', () => {
  cy.visit('/')
  cy.url().should('contain', 'baz.com')
})
```

<Alert type="info">

A future version of Cypress will allow the use of nested `cy.origin()` calls.

> </Alert>

Reviewer 6: Should we remove this notice? When we do the work, we can announce it but this will cause people to ask for when this will be supported.

    cy.visit('/login')
    cy.contains('Username').find('input').type(username)
    cy.contains('Password').find('input').type(password)
    cy.get('button').contains('Login').click()
  })
  // Confirm we're back at the primary origin before continuing
  cy.url().should('contain', '/home')
})
```

Having to go through an entire login flow before every test is not very
performant. Up until now you could get around this problem by putting login code
in the first test of your file, then performing subsequent tests reusing the
same session.

However, once the `experimentalSessionAndOrigin` flag is activated this is no
longer possible, as all session state is now cleared between tests. So to avoid
this overhead we have added the [`cy.session()`](/api/commands/session) command,
which allows you to easily cache session information and reuse it across tests.
So now let's enhance our custom login command with `cy.session()` for a complete
syndicated login flow with session caching and validation. No mocking, no
workarounds, no third-party plugins!

```js
Cypress.Commands.add('login', (username, password) => {
  const args = { username, password }
  cy.session(

>     // Username & password can be used as the cache key too

Reviewer 6: We probably shouldn't recommend including password in the session ID since this could leak private/secret information in the Command Log & in test recordings.

    args,
    () => {
      cy.origin('my-auth.com', { args }, ({ username, password }) => {
        cy.visit('/login')
        cy.contains('Username').find('input').type(username)
        cy.contains('Password').find('input').type(password)
        cy.get('button').contains('Login').click()
      })
      cy.url().should('contain', '/home')
    },
    {
      validate() {
        cy.request('/api/user').its('status').should('eq', 200)
      },
    }
  )
})
```

## Notes

### Migrating existing tests

Enabling the `experimentalSessionAndOrigin` flag makes the test-runner work
slightly differently, and some test suites that rely on the existing behaviour
may have to be updated. The most important of these changes is **test
isolation**. This means that after every test, the current page is reset to
`about:blank` and all active session data

> (cookies, `localStorage` and `sessionStorage`) across all domains are cleared.

Reviewer 3: cookies, localStorage and sessionStorage  are cleared between tests.

This change is opt-in for now, but will be standardized in a future major

> release of Cypress, so eventually all tests will need to be isolated.

Reviewer 6: Should we link/mention this is an existing Best Practice we recommend? https://docs.cypress.io/guides/references/best-practices#Organizing-Tests-Logging-In-Controlling-State

Reviewer 6: https://docs.cypress.io/guides/references/best-practices#Having-tests-rely-on-the-state-of-previous-tests

Reviewer 6: nevermind, reading further these are mentioned.

Before this change, it was possible to write tests such that you could, for

> example, login to a CMS in the first test, change some content in the second

Reviewer 7: ```suggestion
example, log in to a CMS in the first test, change some content in the second
```

`args` option:

**<Icon name="check-circle" color="green"></Icon> Correct Usage**

```js
const foo = 1
cy.origin('somesite.com', { args: { foo } }, ({ foo }) => {
  cy.visit('/')
  // Now it will pass
  cy.get('input').type(foo)
})
```

Cypress uses
[the structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm)
to transfer the `args` option to the secondary origin. This introduces a number
of
[restrictions on the data which may be passed](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm#things_that_dont_work_with_structured_clone).

### Callback restrictions

Because of the way in which the callback is transmitted and executed, there are
certain limitations on what code may be run inside it. In particular, the
following Cypress commands will throw errors if used in the callback:

- `cy.origin()`
- [`cy.intercept()`](/api/commands/intercept)
- [`cy.session()`](/api/commands/session)
- [`cy.server()`](/api/commands/server)
- [`cy.route()`](/api/commands/route)
- [`Cypress.Cookies.preserveOnce()`](/api/cypress-api/cookies)

> 

Reviewer 6: Missing a couple: https://github.com/cypress-io/cypress/blob/feature-multidomain/packages/driver/src/multi-domain/unsupported_apis.ts
Cypress.Server.get()
Cypress.Server.set()
Cypress.session.get()
Cypress.session.set()

Reviewer 7: `get` and `set` aren't the methods. That's us using setters and getters to catch any possible use of `Cypress.Server` and `Cypress.session`.

The user-facing methods are:

- Cypress.Server.defaults()
- Cypress.session.clearAllSavedSessions()

It is also currently not possible to use
[`require()`](https://nodejs.org/en/knowledge/getting-started/what-is-require/)
or
[dynamic `import()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#dynamic_imports)
within the callback. Because of this limitation, it cannot use
[npm](https://www.npmjs.com/) packages or other third-party libraries inside the
callback, as there is no mechanism to reference them. This functionality will be
provided in a future version of Cypress.

While third-party packages are strictly unavailable, it is possible to reuse
your **own** code between `cy.origin()` callbacks. The workaround is to create a
custom Cypress command within the secondary origin in a `before` block:

```js
before(() => {
  cy.origin('somesite.com', () => {
    Cypress.Commands.add('clickLink', (label) => {
      cy.get('a').contains(label).click()
    })
  })
})

it('clicks the secondary origin link', () => {

>   cy.origin('somesite.com', () => {
>     cy.clickLink('Click Me')

Reviewer 3: ```suggestion
  cy.origin('somesite.com', () => {
    cy.visit('https://www.somesite.com/page')
    cy.clickLink('Click Me')
```
